lib/archethic/contracts/loader.ex

```diff
@@ -42,90 +61,163 @@ defmodule Archethic.Contracts.Loader do
+         authorized_nodes
+       )
+       when code != "" do
+    with true <- Election.chain_storage_node?(genesis_address, node_key, authorized_nodes),
+         {:ok, contract} <- Contract.from_transaction(tx),
+         true <- contract_contains_trigger?(contract) do
+      if Worker.exists?(genesis_address),
+        do: Worker.set_contract(genesis_address, contract),
+        else: Worker.new(genesis_address, contract)
+
+      Logger.info("Smart contract loaded",
         transaction_address: Base.encode16(address),
         transaction_type: type
       )
+    else
+      _ -> stop_contract(genesis_address)
+    end
+  end
+
+  defp handle_contract_chain(_, genesis_address, _, _), do: stop_contract(genesis_address)
+
+  defp handle_contract_call(
+         %Transaction{
+           validation_stamp: %ValidationStamp{
+             recipients: resolved_recipients,
+             protocol_version: protocol_version
+           }
+         },
+         node_key,
+         authorized_nodes,
+         execute_contract?
+       )
+       when length(resolved_recipients) > 0 do
+    resolved_recipients
+    |> resolve_genesis_address(authorized_nodes, protocol_version)
+    |> Enum.each(fn contract_genesis_address ->
+      if Election.chain_storage_node?(contract_genesis_address, node_key, authorized_nodes) do
+        with true <- execute_contract?,
+             {trigger_tx, recipient} <- get_next_call(contract_genesis_address),
+             :ok <- request_worker_lock(contract_genesis_address) do
+          Worker.execute(contract_genesis_address, trigger_tx, recipient)
+        end
+      end
     end)
   end
 
-  @doc """
-  Termine a contract execution
-  """
-  @spec stop_contract(binary()) :: :ok
-  def stop_contract(address) when is_binary(address) do
-    case Registry.lookup(ContractRegistry, address) do
-      [{pid, _}] ->
-        DynamicSupervisor.terminate_child(ContractSupervisor, pid)
-        TransactionLookup.clear_contract_transactions(address)
-        # TransactionChain.clear_pending_transactions(address)
-        Logger.info("Stop smart contract at #{Base.encode16(address)}")
-
-      _ ->
-        :ok
-    end
+  defp handle_contract_call(_, _, _, _), do: :ok
+
+  defp resolve_genesis_address(recipients, authorized_nodes, protocol_version)
+       when protocol_version <= 7 do
+    Task.Supervisor.async_stream(
+      Archethic.TaskSupervisor,
+      recipients,
+      fn address ->
+        nodes = Election.chain_storage_nodes(address, authorized_nodes)
+        TransactionChain.fetch_genesis_address(address, nodes)
+      end,
+      on_timeout: :kill_task,
+      ordered: true
+    )
+    |> Enum.zip(recipients)
+    |> Enum.map(fn
+      {{:ok, {:ok, genesis_address}}, recipient} -> {recipient, genesis_address}
+      {_, recipient} -> {recipient, recipient}
+    end)
+  end
+
+  defp resolve_genesis_address(recipients, _, _), do: recipients
+
+  defp contract_contains_trigger?(%Contract{triggers: triggers}) do
```

Comment:
I think it can be interesting to add this function to the Contract module.
Also this will be easier to unit test